timemachine/fe/absolute_hydration.py

```diff
@@ -0,0 +1,153 @@
+"""Absolute hydration free energies"""
+
+from functools import partial
+from typing import List, Sequence
+
+import numpy as np
+from numpy.typing import NDArray as Array
+
+from timemachine.constants import BOLTZ
+from timemachine.fe import functional
+from timemachine.fe.lambda_schedule import construct_pre_optimized_absolute_lambda_schedule_solvent
+from timemachine.ff import Forcefield
+from timemachine.md import enhanced, moves
+from timemachine.md.smc import conditional_multinomial_resample
+from timemachine.md.states import CoordsVelBox
+
+
+def generate_endstate_samples(
+    num_samples: int,
+    solvent_samples: Sequence[CoordsVelBox],
+    ligand_samples: Sequence[CoordsVelBox],
+    ligand_log_weights: Array,
+    num_ligand_atoms: int,
+) -> List[CoordsVelBox]:
+    """solvent + (noninteracting ligand) sample --> solvent + (vacuum ligand) sample
+
+    Inputs
+    ------
+    * solvent_samples ~ p_noninteracting([x_solvent, x_ligand])
+    * (ligand_samples, ligand_log_weights) ~ p_vacuum(x_ligand)
+        (note: this set of samples is importance-weighted!)
+
+    Processing
+    ----------
+    * resample ligand_samples according to ligand_log_weights
+    * concatenate solvent component from p_noninteracting with ligand from p_vacuum
+
+    Returns
+    -------
+    * list of (coordinates, velocities, box), distributed according to p_noninteracting
+
+    Assumptions
+    -----------
+    * ligand indices: last num_ligand_atoms
+
+    Notes
+    -----
+    * TODO[generality]: refactor to accept two streams of unweighted samples, concatenate them
+        (rather than requiring and discarding ligand component in solvent_samples, performing multinomial resampling)
+    * TODO[logic]: generate solvent samples once independently of ligand
+        (rather than duplicating work of solvent sampling across all ligands)
+    """
+
+    # assume this layout
+    num_total_atoms = len(solvent_samples[0].coords)
+    num_solvent_atoms = num_total_atoms - num_ligand_atoms
+    assert num_solvent_atoms > 0, "Oops, did you really mean num_ligand_atoms >= num_total_atoms?"
+    solvent_idxs = np.arange(0, num_solvent_atoms)
+
+    # sample according to log weights
+    ligand_xvs = enhanced.sample_from_log_weights(ligand_samples, ligand_log_weights, size=num_samples)
+
+    # sample uniformly with replacement
+    solvent_choice_idxs = np.random.choice(len(solvent_samples), size=num_samples)
```

Comment:
nit: might be nice to pass `replace=True` here (even though it's apparently the default) for a more self-documenting style

Reply:
Good call -- added here and at one other site https://github.com/proteneer/timemachine/pull/718/commits/63fb45f3814f9bdf91579c63200bef1060dc643c